presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -593,22 +705,29 @@ else if (!result.getBlocked().isDone()) {
                 }
             }
 
-            for (SqlStageExecution stage : stages.values()) {
+            for (SqlStageExecution stage : stages) {
                 StageState state = stage.getState();
                 if (state != SCHEDULED && state != RUNNING && !state.isDone()) {
                     throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Scheduling is complete, but stage %s is in state %s", stage.getStageId(), state));
                 }
             }
+
+            scheduling.set(false);
```

Comment:
So, `scheduling` variable guard only one thread can enter the previous code section.

It's not immediately clear to me what happens, if one thread is scheduling stages, while another thread is trying to get the next batch of stages to be scheduled.

Reply:
> It's not immediately clear to me what happens, if one thread is scheduling stages, while another thread is trying to get the next batch of stages to be scheduled.

The second thread will check the `scheduling` variable, and if it is set to `true` - it won't start scheduling. Whenever scheduling is finished - the `schedule` method set's the flag back to false, and checks when there's anything to schedule.